mujoco_warp/_src/ray.py

```diff
@@ -0,0 +1,588 @@
+  """Returns the distance at which a ray intersects with a sphere."""
+  a = wp.dot(vec, vec)
+  b = wp.dot(vec, pnt)
+  c = wp.dot(pnt, pnt) - size[0] * size[0]
+
+  solutions = _ray_quad(a, b, c)
+  x0 = solutions[0]
+  x1 = solutions[1]
+  x = wp.where(wp.isinf(x0), x1, x0)
+
+  return_id = wp.where(wp.isinf(x), -1, geom_id)
+  return DistanceWithId(x, return_id)
+
+
+@wp.func
+def _ray_capsule(
+  size: wp.vec3,
+  pnt: wp.vec3,
+  vec: wp.vec3,
+  geom_id: int,
+) -> DistanceWithId:
+  """Returns the distance at which a ray intersects with a capsule."""
+
+  # cylinder round side: (x*lvec+lpnt)'*(x*lvec+lpnt) = size[0]*size[0]
+  # For a capsule, we only care about the x,y components when checking cylinder intersection
+  # since the z component is handled separately with the caps
+  a = wp.dot(wp.vec2(vec[0], vec[1]), wp.vec2(vec[0], vec[1]))
+  b = wp.dot(wp.vec2(pnt[0], pnt[1]), wp.vec2(vec[0], vec[1]))
+  c = wp.dot(wp.vec2(pnt[0], pnt[1]), wp.vec2(pnt[0], pnt[1])) - size[0] * size[0]
+
+  # solve a*x^2 + 2*b*x + c = 0
+  solutions = _ray_quad(a, b, c)
+  x0 = solutions[0]
+  x1 = solutions[1]
+  x = wp.where(wp.isinf(x0), x1, x0)
+
+  # make sure round solution is between flat sides
+  x = wp.where(wp.abs(pnt[2] + x * vec[2]) <= size[1], x, wp.inf)
+
+  # top cap
+  dif = pnt - wp.vec3(0.0, 0.0, size[1])
+  solutions = _ray_quad(
+    wp.dot(vec, vec), wp.dot(vec, dif), wp.dot(dif, dif) - size[0] * size[0]
+  )
+  x0 = solutions[0]
+  x1 = solutions[1]
+  # accept only top half of sphere
+  x = wp.where((pnt[2] + x0 * vec[2] >= size[1]) and (x0 < x), x0, x)
+  x = wp.where((pnt[2] + x1 * vec[2] >= size[1]) and (x1 < x), x1, x)
+
+  # bottom cap
+  dif = pnt + wp.vec3(0.0, 0.0, size[1])
+  solutions = _ray_quad(
+    wp.dot(vec, vec), wp.dot(vec, dif), wp.dot(dif, dif) - size[0] * size[0]
+  )
+  x0 = solutions[0]
+  x1 = solutions[1]
+  # accept only bottom half of sphere
+  x = wp.where((pnt[2] + x0 * vec[2] <= -size[1]) and (x0 < x), x0, x)
+  x = wp.where((pnt[2] + x1 * vec[2] <= -size[1]) and (x1 < x), x1, x)
+
+  return_id = wp.where(wp.isinf(x), -1, geom_id)
+  return DistanceWithId(x, return_id)
+
+
+@wp.func
+def _ray_ellipsoid(
+  size: wp.vec3,
+  pnt: wp.vec3,
+  vec: wp.vec3,
+  geom_id: int,
+) -> DistanceWithId:
+  """Returns the distance at which a ray intersects with an ellipsoid."""
+
+  # invert size^2
+  s = wp.vec3(
+    1.0 / (size[0] * size[0]), 1.0 / (size[1] * size[1]), 1.0 / (size[2] * size[2])
+  )
+
+  # (x*lvec+lpnt)' * diag(1/size^2) * (x*lvec+lpnt) = 1
+  svec = wp.vec3(s[0] * vec[0], s[1] * vec[1], s[2] * vec[2])
+  a = wp.dot(svec, vec)
+  b = wp.dot(svec, pnt)
+  c = wp.dot(wp.vec3(s[0] * pnt[0], s[1] * pnt[1], s[2] * pnt[2]), pnt) - 1.0
+
+  # solve a*x^2 + 2*b*x + c = 0
+  solutions = _ray_quad(a, b, c)
+  x0 = solutions[0]
+  x1 = solutions[1]
+  x = wp.where(wp.isinf(x0), x1, x0)
+
+  return_id = wp.where(wp.isinf(x), -1, geom_id)
+  return DistanceWithId(x, return_id)
+
+
+@wp.func
+def _ray_box(
```

Comment:
this function can probably follow the mujoco reference more closely
https://github.com/google-deepmind/mujoco/blob/f317bd17c3494b954982de7bcf14030978d5013b/src/engine/engine_ray.c#L387